File: summary/ts/25_union_find.ts
class UnionFind {
    private parent: number[];
    private rank: number[];

    constructor(size: number) {
        this.parent = new Array(size).fill(0).map((_, index) => index);
        this.rank = new Array(size).fill(0);
    }

    public find(x: number): number {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    public union(x: number, y: number): void {
        const rootX = this.find(x);
        const rootY = this.find(y);

        if (rootX !== rootY) {
            if (this.rank[rootX] > this.rank[rootY]) {
                this.parent[rootY] = rootX;
            } else if (this.rank[rootX] < this.rank[rootY]) {
                this.parent[rootX] = rootY;
            } else {
                this.parent[rootY] = rootX;
                this.rank[rootX] += 1;
            }
        }
    }

    public connected(x: number, y: number): boolean {
        return this.find(x) === this.find(y);
    }
}

// Example usage
const uf = new UnionFind(10);
uf.union(1, 2);
uf.union(2, 3);
console.log(uf.connected(1, 3)); // true
console.log(uf.connected(1, 4)); // false

File: summary/dart/25_union_find.dart
class UnionFind {
  List<int> parent;
  List<int> rank;

  UnionFind(int size)
      : parent = List<int>.generate(size, (i) => i),
        rank = List<int>.filled(size, 0);

  int find(int x) {
    if (parent[x] != x) {
      parent[x] = find(parent[x]); // パス圧縮
    }
    return parent[x];
  }

  void union(int x, int y) {
    int rootX = find(x);
    int rootY = find(y);

    if (rootX != rootY) {
      // ランクによる統合
      if (rank[rootX] > rank[rootY]) {
        parent[rootY] = rootX;
      } else if (rank[rootX] < rank[rootY]) {
        parent[rootX] = rootY;
      } else {
        parent[rootY] = rootX;
        rank[rootX]++;
      }
    }
  }
}

void main() {
  UnionFind uf = UnionFind(10);

  uf.union(1, 2);
  uf.union(2, 3);
  uf.union(4, 5);
  uf.union(6, 7);
  uf.union(5, 6);

  print(uf.find(1)); // 出力例: 1
  print(uf.find(2)); // 出力例: 1
  print(uf.find(3)); // 出力例: 1
  print(uf.find(4)); // 出力例: 4
  print(uf.find(5)); // 出力例: 4
  print(uf.find(6)); // 出力例: 4
  print(uf.find(7)); // 出力例: 4
  print(uf.find(8)); // 出力例: 8
}
